File: app/src/main/java/info/nightscout/androidaps/Services/AlarmSoundService.java
package info.nightscout.androidaps.Services;

import android.app.Service;
import android.content.Intent;
import android.content.res.AssetFileDescriptor;
import android.media.MediaPlayer;
import android.os.IBinder;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;

public class AlarmSoundService extends Service {
    private static Logger log = LoggerFactory.getLogger(AlarmSoundService.class);

    MediaPlayer player;
    int resourceId = R.raw.error;

    public AlarmSoundService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void onCreate() {
        super.onCreate();
        log.debug("onCreate");
    }

    public int onStartCommand(Intent intent, int flags, int startId) {
        if (player != null && player.isPlaying())
            player.stop();
        log.debug("onStartCommand");
        if (intent != null && intent.hasExtra("soundid"))
            resourceId = intent.getIntExtra("soundid", R.raw.error);

        player = new MediaPlayer();
        AssetFileDescriptor afd = MainApp.sResources.openRawResourceFd(resourceId);
        if (afd == null)
            return START_STICKY;
        try {
            player.setDataSource(afd.getFileDescriptor(), afd.getStartOffset(), afd.getLength());
            afd.close();
        } catch (IOException e) {
            log.error("Unhandled exception", e);
        }
        player.setLooping(true); // Set looping
        player.setVolume(100, 100);

        try {
            player.prepare();
            player.start();
        } catch (IOException e) {
            log.error("Unhandled exception", e);
        }

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        player.stop();
        player.release();
    }
}

File: core/src/main/java/info/nightscout/androidaps/services/AlarmSoundService.kt
package info.nightscout.androidaps.services

import android.content.Context
import android.content.Intent
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.IBinder
import dagger.android.DaggerService
import info.nightscout.androidaps.core.R
import info.nightscout.androidaps.interfaces.NotificationHolderInterface
import info.nightscout.androidaps.logging.AAPSLogger
import info.nightscout.androidaps.logging.LTag
import info.nightscout.androidaps.utils.resources.ResourceHelper
import javax.inject.Inject

class AlarmSoundService : DaggerService() {
    @Inject lateinit var aapsLogger: AAPSLogger
    @Inject lateinit var resourceHelper: ResourceHelper
    @Inject lateinit var notificationHolder: NotificationHolderInterface

    private var player: MediaPlayer? = null
    private var resourceId = R.raw.error

    override fun onBind(intent: Intent): IBinder? = null

    override fun onCreate() {
        super.onCreate()
        aapsLogger.debug(LTag.CORE, "onCreate")
        startForeground(notificationHolder.notificationID, notificationHolder.notification)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        startForeground(notificationHolder.notificationID, notificationHolder.notification)

        player?.let { if (it.isPlaying) it.stop() }

        aapsLogger.debug(LTag.CORE, "onStartCommand")
        if (intent.hasExtra("soundid")) resourceId = intent.getIntExtra("soundid", R.raw.error)
        player = MediaPlayer()
        try {
            val afd = resourceHelper.openRawResourceFd(resourceId) ?: return START_STICKY
            player?.setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
            afd.close()
            player?.isLooping = true // Set looping
            val manager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
            if (!manager.isMusicActive) {
                player?.setVolume(100f, 100f)
            }
            player?.prepare()
            player?.start()
        } catch (e: Exception) {
            aapsLogger.error("Unhandled exception", e)
        }
        return START_STICKY
    }

    override fun onDestroy() {
        player?.stop()
        player?.release()
        aapsLogger.debug(LTag.CORE, "onDestroy")
    }
}